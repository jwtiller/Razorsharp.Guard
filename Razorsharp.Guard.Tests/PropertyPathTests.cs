using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc.Abstractions;
using Microsoft.AspNetCore.Mvc.Filters;
using Microsoft.AspNetCore.Mvc.ModelBinding;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Routing;
using System.Threading.Tasks.Sources;
using Microsoft.Extensions.Logging.Abstractions;
using Snapshooter.NUnit;
using Razorsharp.Guard.Entities;

namespace Razorsharp.Guard.Tests
{
    [TestFixture]
    public class PropertyPathTests
    {
        private List<string> _audits = new();
        private ClassificationFilter CreateFilter() => new ClassificationFilter(
            new GuardOptions() { 
            GuardMode = GuardMode.ThrowExceptionAndCallback,
            Callback = (logger,context,evt) => {
                var details = string.Join(',',evt.Classifications.Select(c => $"{c.Type}[{c.SensitivityLevel}={c.Reason}]"));
                _audits.Add($"Sensitive data accessed: {details}"); }
            }, new NullLogger<ClassificationFilter>());

        private ResultExecutingContext CreateContext(object dto)
        {
            var httpContext = new DefaultHttpContext();
            var actionContext = new ActionContext(
                httpContext,
                new RouteData(),
                new ActionDescriptor(),
                new ModelStateDictionary()
            );

            return new ResultExecutingContext(
                actionContext,
                new List<IFilterMetadata>(),
                new ObjectResult(dto),
                controller: null
            );
        }

        [Public]
        public class ProductDTO
        {
            [Public]
            public string Name { get; set; } = string.Empty;

            [Public]
            public Manufacturer Manufacturer { get; set; } = new();
        }

        [Internal("Contains supplier contact information.")]
        public class Manufacturer
        {
            [Public]
            public string CompanyName { get; set; } = string.Empty;

            [Restricted("Supplier address may reveal internal logistics location.")]
            public FactoryAddress FactoryAddress { get; set; } = new();
        }

        [Restricted("Physical factory address classified as internal data.")]
        public class FactoryAddress
        {
            [Confidential("Street information is not for public disclosure.")]
            public string Street { get; set; } = string.Empty;

            [Public]
            public string City { get; set; } = string.Empty;
        }


        [Test]
        public void Propertyname()
        {
            var filter = CreateFilter();
            var context = CreateContext(new ProductDTO());

            Assert.Throws<RazorsharpGuardException>(() => filter.OnResultExecuting(context));

            filter.Classifications.MatchSnapshot();
        }
    }
}
