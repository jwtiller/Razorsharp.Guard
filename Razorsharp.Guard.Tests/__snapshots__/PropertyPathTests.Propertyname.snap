[
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO",
    "ParentType": null,
    "Reason": null,
    "SensitivityLevel": "Public",
    "Depth": 0,
    "AttributeLevel": "Class"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO.Name",
    "ParentType": null,
    "Reason": null,
    "SensitivityLevel": "Public",
    "Depth": 0,
    "AttributeLevel": "Property"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO.Manufacturer",
    "ParentType": null,
    "Reason": null,
    "SensitivityLevel": "Public",
    "Depth": 0,
    "AttributeLevel": "Property"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO",
    "Reason": "Contains supplier contact information.",
    "SensitivityLevel": "Internal",
    "Depth": 1,
    "AttributeLevel": "Class"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer.CompanyName",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO",
    "Reason": null,
    "SensitivityLevel": "Public",
    "Depth": 1,
    "AttributeLevel": "Property"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer.FactoryAddress",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+ProductDTO",
    "Reason": "Supplier address may reveal internal logistics location.",
    "SensitivityLevel": "Restricted",
    "Depth": 1,
    "AttributeLevel": "Property"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+FactoryAddress",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer",
    "Reason": "Physical factory address classified as internal data.",
    "SensitivityLevel": "Restricted",
    "Depth": 2,
    "AttributeLevel": "Class"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+FactoryAddress.Street",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer",
    "Reason": "Street information is not for public disclosure.",
    "SensitivityLevel": "Confidential",
    "Depth": 2,
    "AttributeLevel": "Property"
  },
  {
    "Type": "Razorsharp.Guard.Tests.PropertyPathTests+FactoryAddress.City",
    "ParentType": "Razorsharp.Guard.Tests.PropertyPathTests+Manufacturer",
    "Reason": null,
    "SensitivityLevel": "Public",
    "Depth": 2,
    "AttributeLevel": "Property"
  }
]
